source/core/Native.cs

```diff
@@ -0,0 +1,711 @@
+            {
+                return new InputArgument(new string(value));
+            }
+
+            // Pointer types
+            public static implicit operator InputArgument(IntPtr value)
+            {
+                return new InputArgument(value);
+            }
+            public static unsafe implicit operator InputArgument(void* value)
+            {
+                return new InputArgument(new IntPtr(value));
+            }
+
+            public static implicit operator InputArgument(OutputArgument value)
+            {
+                return new InputArgument(value._storage);
+            }
+        }
+
+        public class OutputArgument
+        {
+            internal IntPtr _storage = Marshal.AllocCoTaskMem(24);
+            private bool disposed;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="OutputArgument"/> class for natives that output data into pointers.
+            /// </summary>
+            public OutputArgument()
+            {
+            }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="OutputArgument"/> class with an initial value for natives that require the pointer to data instead of the actual data.
+            /// </summary>
+            /// <param name="initvalue">The value to set the data of this <see cref="OutputArgument"/> to.</param>
+            public OutputArgument(object value) : this()
+            {
+                unsafe
+                {
+                    *(ulong*)(_storage) = Function.ObjectToNative(value);
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+            protected virtual void Dispose(bool disposing)
+            {
+                if (!disposed)
+                {
+                    if (disposing)
+                    {
+                    }
+                    Marshal.FreeCoTaskMem(_storage);
+                    disposed = true;
+                }
+            }
+            ~OutputArgument()
+            {
+                Dispose(false);
+            }
+
+            /// <summary>
+            /// Gets the value of data stored in this <see cref="OutputArgument"/>.
+            /// </summary>
+            public T GetResult<T>()
+            {
+                unsafe
+                {
+                    return (T)(Function.ObjectFromNative(typeof(T), (ulong*)(_storage.ToPointer())));
+                }
+            }
+        }
+
+        public sealed class Function
+        {
+            internal unsafe struct NativeTask : IScriptTask
+            {
+                internal ulong _hash;
+                internal ulong* _result;
+                internal InputArgument[] _arguments;
+
+                public void Run()
+                {
+					Wrapper.NativeInit(_hash);
```

Comment:
Why use a separate wrapper library? Wouldn't it be much simpler to simply use P/Invoke?
```c#
[DllImport("ScriptHookV.dll", CallingConvention = CallingConvention.Cdecl, EntryPoint = "nativeInit")]
static extern void NativeInit(ulong hash);
[DllImport("ScriptHookV.dll", CallingConvention = CallingConvention.Cdecl, EntryPoint = "nativePush64")]
static extern void NativePush64(ulong val);
[DllImport("ScriptHookV.dll", CallingConvention = CallingConvention.Cdecl, EntryPoint = "nativeCall")]
static extern ulong* NativeCall();
```
If that doesn't work (since ScriptHookV.dll uses mangled C++ exports and not C exports), we can use the ordinal index instead of the function name to look the exports up:
```c#
[DllImport("ScriptHookV.dll", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true, EntryPoint = "#9")]
static extern void NativeInit(ulong hash); // ?nativeInit@@YAX_K@Z
[DllImport("ScriptHookV.dll", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true, EntryPoint = "#10")]
static extern void NativePush64(ulong val); // ?nativePush64@@YAX_K@Z
[DllImport("ScriptHookV.dll", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true, EntryPoint = "#8")]
static extern ulong* NativeCall(); // ?nativeCall@@YAPEA_KXZ
```

Reply:
Your concern was right, I confirmed the former P/Invoke lines crash GTA.